### PR TITLE
Add Featuretools Component

### LINKED ISSUE #470: Add featuretools component
A featuretools component could help us handle different types of variable types such as datetimes, text, or zip codes. 

In order to do this, we'd need to come up with an approach for passing along variable type metadata about columns in the input data frame in the pipeline

### PR DESCRIPTION
**Pull request by bchen1116**
fix #470 

Adding featuretools component to evalml.

Quip doc [here](https://alteryx.quip.com/u7keAyIAtXOB/Featuretools-Component)
Perf tests [here](https://alteryx.quip.com/bHIEAzPegZ2f/Featuretools-Component-Perf-Tests-single-dataframesdatatables)
Docs [here](https://evalml.alteryx.com/en/bc_470_featuretools/generated/evalml.pipelines.components.DFSTransformer.html#evalml.pipelines.components.DFSTransformer)

The perf test are run on AutoMLSearch cv folds, not holdout data. The perf tests include more info.

### CODE REVIEW

File: evalml/pipelines/components/transformers/preprocessing/featuretools.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        super().__init__(parameters=parameters,
+                         random_state=random_state)
+
+    def _make_entity_set(self, X):
+        """helper method that creates and returns the entity set given the datatable X
+        """
+        if 'index' not in X.columns:
+            es = self._ft_es.entity_from_dataframe(entity_id="X", dataframe=X, index='index', make_index=True)
+        else:
+            es = self._ft_es.entity_from_dataframe(entity_id="X", dataframe=X, index='index')
```
Comment (bchen1116) [inline evalml/pipelines/components/transformers/preprocessing/featuretools.py:36]: `Featuretools` automatically uses the first column as the index if an `index` column isn't provided.

File: evalml/pipelines/components/transformers/preprocessing/featuretools.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    def __init__(self, max_depth=2, random_state=0, **kwargs):
+        """Allows for featuretools to be used in EvalML
+
+        Arguments:
+            max_depth (int): The max allowed depth of features. Defaults to 2
+            random_state (int, np.random.RandomState): seed for the random number generator
+        """
+        parameters = {"max_depth": max_depth}
+        self._ft_es = EntitySet()
```
Comment (freddyaboulton) [inline evalml/pipelines/components/transformers/preprocessing/featuretools.py:23]: Why not define this within `_make_entity_set` and set `_make_entity_set` as static?

File: evalml/tests/component_tests/test_featuretools.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        X_pd = pd.DataFrame(X)
+        feature.fit(X_pd, y)
+        feature.transform(X_pd)
+
+        X_ww = ww.DataTable(X_pd)
+        feature.fit(X_ww)
+        feature.transform(X_ww)
+
+
+def test_transform(X_y_binary, X_y_multi, X_y_regression):
```
Comment (freddyaboulton) [inline evalml/tests/component_tests/test_featuretools.py:67]: I think it's ok to keep the tests as-is but I think we can condense this one and `test_input_types` with `@parametrize`!

**Review by freddyaboulton — APPROVED** [review]
@bchen1116 This is sweet! I agree with your analysis that we should hold off on adding this to `_make_preprocessing_components`.

I think the implementation looks good. My main comment is about letting users specify the index column in the entity set to avoid name collisions with columns already named `index` that aren't intended to be used as indices.

File: evalml/pipelines/components/transformers/preprocessing/featuretools.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        self.max_depth = max_depth
+        self.features = None
+        parameters.update(kwargs)
+        super().__init__(parameters=parameters,
+                         random_state=random_state)
+
+    def _make_entity_set(self, X):
```
[the remark this answers is not in the document]
Comment (bchen1116) [reply evalml/pipelines/components/transformers/preprocessing/featuretools.py:30]: Great point! I'll add that in

**Comment by bchen1116** [comment]
Perf test [doc](https://alteryx.quip.com/bHIEAzPegZ2f/Featuretools-Component-Perf-Tests-single-dataframesdatatables) has been updated with new results after Looking Glass's recent [bug fix](https://github.com/FeatureLabs/evalml_looking_glass/pull/183)

**Review by jeremyliweishih — APPROVED** [review]
Looks good to me Bryan. I agree with you on not adding this to AutoML in the interim and wait for a deeper implementation. I'm excited to see the experimentation results then!

File: evalml/pipelines/components/transformers/preprocessing/featuretools.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
+from featuretools import EntitySet, calculate_feature_matrix, dfs
+
+from evalml.pipelines.components.transformers.transformer import Transformer
+from evalml.utils.gen_utils import (
+    _convert_to_woodwork_structure,
+    _convert_woodwork_types_wrapper
+)
+
+
+class DFSTransformer(Transformer):
```
Comment (dsherry) [inline evalml/pipelines/components/transformers/preprocessing/featuretools.py:10]: @bchen1116 any reason not to call this `FeaturetoolsTransformer`? I'm not sure people will know what "DFS" means, or will think its a reference to depth-first search 😆

File: evalml/pipelines/components/transformers/preprocessing/featuretools.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+        Arguments:
+            X (ww.DataTable, pd.DataFrame or np.ndarray): The input training data to transform. Has shape [n_samples, n_features]
+        """
+        X = _convert_to_woodwork_structure(X)
+        X = _convert_woodwork_types_wrapper(X.to_dataframe())
+        X.columns = X.columns.astype(str)
+        es = self._make_entity_set(X)
+        feature_matrix = calculate_feature_matrix(features=self.features, entityset=es)
+        return feature_matrix
```
Comment (dsherry) [inline evalml/pipelines/components/transformers/preprocessing/featuretools.py:70]: Very cool!

**Review by dsherry — COMMENTED** [review]
@bchen1116 wonderful! Left one comment about component name, impl and tests look great.

File: evalml/pipelines/components/transformers/preprocessing/featuretools.py (hunk 5 above)
Comment (bchen1116) [reply evalml/pipelines/components/transformers/preprocessing/featuretools.py:10]: Ah that's a fair point. I can change that in the current PR!

**Pull request merged**